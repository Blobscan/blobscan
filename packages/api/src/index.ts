import type { inferRouterInputs, inferRouterOutputs } from "@trpc/server";

import type { AppRouter } from "./app-router";

export { appRouter, type AppRouter } from "./app-router";
export { createTRPCContext, createTRPCInnerContext } from "./context";
export type {
  TRPCInnerContext as TRCInnerContext,
  TRPCContext,
} from "./context";

export * from "@trpc/server/adapters/express";

export { env } from "./env";
export { gracefulShutdown } from "./graceful-shutdown";
export { metricsHandler } from "./instrumentation";
export type {
  BlobStorage,
  Rollup,
  Metric,
  MetricHistogram,
  MetricHistogramBucket,
  Metrics,
  MetricsClient,
} from "@blobscan/db";

/**
 * Inference helpers for input types
 * @example type HelloInput = RouterInputs['example']['hello']
 **/
export type RouterInputs = inferRouterInputs<AppRouter>;

/**
 * Inference helpers for output types
 * @example type HelloOutput = RouterOutputs['example']['hello']
 **/
export type RouterOutputs = inferRouterOutputs<AppRouter>;
