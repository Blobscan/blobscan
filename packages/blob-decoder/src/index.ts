export * from "./blob-decoder";
export * from "./types";
