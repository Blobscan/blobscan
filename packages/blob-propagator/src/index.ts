export type { EnvVars as Environment } from "./env";

export { env } from "./env";
export { BlobPropagator } from "./BlobPropagator";
export { blobPropagator } from "./blob-propagator";
export * from "./types";
export {
  createBlobPropagationFlowJob,
  createBlobStorageJob,
  FINALIZER_WORKER_NAME,
  STORAGE_WORKER_NAMES,
  buildJobId,
} from "./utils";
