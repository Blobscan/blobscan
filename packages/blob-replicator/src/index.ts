export {
  isBlobReplicationAvailable,
  queueBlobForReplication,
  queueBlobsForReplication,
} from "./replication";
export * from "./types";
export {
  setUpBlobReplicationWorkers,
  tearDownBlobReplicationWorkers,
} from "./workers";
