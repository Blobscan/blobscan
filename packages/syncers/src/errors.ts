export class ErrorException extends Error {
  constructor(message: string, cause?: unknown) {
    super(message, {
      cause,
    });

    this.name = this.constructor.name;
  }
}

export class SyncerError extends ErrorException {
  constructor(syncerName: string, message: string, cause: unknown) {
    super(`Syncer "${syncerName}" failed: ${message}`, cause);
  }
}
