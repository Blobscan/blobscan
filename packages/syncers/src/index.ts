export { BaseSyncer } from "./BaseSyncer";
export { createRedisConnection } from "./utils";

export { syncSwarmStamp } from "./syncers/SwarmStampSyncer";
export { aggregateDailyStats } from "./syncers/DailyStatsSyncer";
export { aggregateOverallStats } from "./syncers/OverallStatsSyncer";
