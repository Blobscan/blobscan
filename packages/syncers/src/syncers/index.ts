export { DailyStatsSyncer } from "./DailyStatsSyncer";
export type { DailyStatsSyncerConfig } from "./DailyStatsSyncer";
export { OverallStatsSyncer } from "./OverallStatsSyncer";
export type { OverallStatsSyncerConfig } from "./OverallStatsSyncer";
