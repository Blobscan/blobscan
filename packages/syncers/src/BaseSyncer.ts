import { Queue, Worker } from "bullmq";
import type { Redis } from "ioredis";

import { createModuleLogger } from "@blobscan/logger";
import type { Logger } from "@blobscan/logger";

import { ErrorException, SyncerError } from "./errors";

export type BaseSyncerConfig = {
  name: string;
  connection: Redis;
  cronPattern: string;
  syncerFn: () => Promise<void>;
};

export class BaseSyncer {
  private name: string;
  private cronPattern: string;
  private logger: Logger;

  public readonly worker: Worker;
  public readonly queue: Queue;

  constructor({ name, cronPattern, connection, syncerFn }: BaseSyncerConfig) {
    this.name = `${name}-syncer`;
    this.cronPattern = cronPattern;
    this.logger = createModuleLogger(this.name);

    this.queue = new Queue(this.name, { connection });
    this.worker = new Worker(this.queue.name, syncerFn, { connection });

    this.queue.on("error", (error) => {
      this.logger.error(new ErrorException("A queue error occurred", error));
    });

    this.worker.on("failed", (_, error) => {
      this.logger.error(new ErrorException("A worker error occurred", error));
    });
  }

  async start() {
    try {
      const repeatableJob = await this.queue.add(`${this.name}-job`, null, {
        repeat: {
          pattern: this.cronPattern,
        },
      });

      this.logger.info("Syncer started successfully");

      return repeatableJob;
    } catch (error) {
      throw new SyncerError(
        this.name,
        "An error occurred when starting syncer",
        error
      );
    }
  }

  async close(): Promise<void> {
    const results = await Promise.allSettled([
      this.queue.obliterate({ force: true }),
      this.queue.removeAllListeners().close(),
      this.worker.removeAllListeners().close(true),
    ]);

    for (const result of results) {
      if (result.status === "rejected") {
        throw new SyncerError(
          this.name,
          "An error occurred when performing closing operation",
          result.reason
        );
      }
    }

    this.logger.info("Syncer closed successfully");
  }
}
