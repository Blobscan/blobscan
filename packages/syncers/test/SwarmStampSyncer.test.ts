import { http, HttpResponse } from "msw";
import { setupServer } from "msw/node";
import { beforeAll, beforeEach, describe, expect, it } from "vitest";

import type { BlobStoragesState } from "@blobscan/db";
import { prisma } from "@blobscan/db";
import { fixtures, testValidError } from "@blobscan/test";

import { syncSwarmStamp } from "../src";

const BEE_ENDPOINT = process.env.BEE_ENDPOINT ?? "http://localhost:1633";

describe("SwarmStampSyncer", () => {
  const expectedBatchId = fixtures.blobStoragesState[0]?.swarmDataId as string;
  const expectedBatchTTL = 1000;

  beforeAll(() => {
    const baseUrl = `${BEE_ENDPOINT}/stamps`;
    const server = setupServer(
      ...[
        http.get(`${baseUrl}/:batchId`, ({ request }) => {
          const batchId = request.url.split("/").pop();

          if (!batchId || batchId.length !== 64) {
            return HttpResponse.json(
              {
                code: 400,
                message: "invalid path params",
                reasons: [
                  {
                    field: "batch_id",
                    error: "odd length hex string",
                  },
                ],
              },
              { status: 400 }
            );
          }

          if (batchId !== expectedBatchId) {
            return HttpResponse.json(
              {
                code: 404,
                message: "issuer does not exist",
              },
              { status: 404 }
            );
          }

          return HttpResponse.json(
            {
              batchID: expectedBatchId,
              batchTTL: expectedBatchTTL,
            },
            {
              status: 200,
            }
          );
        }),
      ]
    );

    server.listen();

    return () => {
      server.close();
    };
  });

  describe("when creating a new swarm batch data row in the db", async () => {
    let blobStorageState: BlobStoragesState | null = null;

    beforeEach(async () => {
      await prisma.blobStoragesState.deleteMany();

      await syncSwarmStamp({
        batchId: process.env.SWARM_BATCH_ID ?? "",
        beeEndpoint: BEE_ENDPOINT,
      }).catch((err) => console.log(err));

      blobStorageState = await prisma.blobStoragesState.findFirst();
    });

    it("should create it with the correct swarm stamp batch ID", async () => {
      expect(blobStorageState?.swarmDataId).toBe(process.env.SWARM_BATCH_ID);
    });

    it("should create it with the correct batch TTL", async () => {
      expect(blobStorageState?.swarmDataTTL).toBe(expectedBatchTTL);
    });
  });

  it("should update the batch TTl", async () => {
    await prisma.blobStoragesState.update({
      data: {
        swarmDataTTL: 99999,
      },
      where: {
        id: 1,
      },
    });

    await syncSwarmStamp({
      batchId: process.env.SWARM_BATCH_ID ?? "",
      beeEndpoint: BEE_ENDPOINT,
    });

    const blobStorageState = await prisma.blobStoragesState.findFirst();

    expect(blobStorageState?.swarmDataTTL).toBe(expectedBatchTTL);
  });

  testValidError(
    "should fail when trying to fetch a non-existing batch",
    async () => {
      await syncSwarmStamp({
        beeEndpoint: BEE_ENDPOINT,
        batchId:
          "6b538866048cfb6e9e1d06805374c51572c11219d2d550c03e6e277366cb0371",
      });
    },
    Error,
    {
      checkCause: true,
    }
  );

  testValidError(
    "should fail when trying to fetch an invalid batch",
    async () => {
      await syncSwarmStamp({
        beeEndpoint: BEE_ENDPOINT,
        batchId: "invalid-batch",
      });
    },
    Error,
    {
      checkCause: true,
    }
  );
});
