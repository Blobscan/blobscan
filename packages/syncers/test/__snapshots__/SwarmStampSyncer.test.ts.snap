// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SwarmStampSyncer > should fail when trying to fetch a non-existing batch 1`] = `"Failed to fetch stamp batch \\"6b538866048cfb6e9e1d06805374c51572c11219d2d550c03e6e277366cb0371\\""`;

exports[`SwarmStampSyncer > should fail when trying to fetch a non-existing batch 2`] = `[SwarmNodeError: issuer does not exist]`;

exports[`SwarmStampSyncer > should fail when trying to fetch an invalid batch 1`] = `"Failed to fetch stamp batch \\"invalid-batch\\""`;

exports[`SwarmStampSyncer > should fail when trying to fetch an invalid batch 2`] = `[SwarmNodeError: invalid path params]`;
